### PR TITLE
Add opt-in native MLX random sampling for no-logprobs decode

### PR DESCRIPTION
**Pull request by CAICAIIs**
Refs #343

## Summary

This PR adds a first, opt-in native MLX random sampling path for paged decode requests that do not request logprobs.

Supported in this first slice:

- temperature
- top_k
- top_p
- min_p
- categorical sampling

The existing greedy `mx.argmax` fast path is unchanged. `sample_from_logits()` remains the single sampling entry point.

## Scope

Native MLX random sampling is enabled only when `VLLM_METAL_NATIVE_RANDOM_SAMPLING=1`.

The new path conservatively falls back to the existing vLLM Torch sampler for:

- logprobs
- seeded generators / `SamplingParams.seed`
- frequency / presence / repetition penalties
- `allowed_token_ids`
- bad words
- active `min_tokens`
- active `logit_bias`
- unknown or unsupported logits processors
- prefill sampling

The env flag defaults to off because the current end-to-end benchmark does not show a stable throughput improvement.

## Correctness Notes

The MLX path applies filters in the same order as the existing sampler path:

1. temperature
2. min_p
3. top_k / top_p
4. categorical sampling

The top-k subset fast path preserves vLLM's boundary-tie semantics: if tokens are tied at the kth threshold, it falls back to the full filtering path so tied candidates are not dropped.

## Validation

Commands were run from the repository root with the project virtualenv.

```bash
.venv-vllm-metal/bin/python -m pytest tests/test_sampling.py -q
# 51 passed, 16 warnings

.venv-vllm-metal/bin/python -m pytest tests/test_sampling.py tests/test_v1_model_runner_generate.py -q
# 86 passed, 16 warnings

.venv-vllm-metal/bin/python -m pytest tests/test_paged_attention.py tests/test_grammar_bitmask.py -q
# 38 passed, 16 warnings

.venv-vllm-metal/bin/python -m pytest tests/test_register.py tests/test_config.py -q
# 21 passed, 2 warnings

.venv-vllm-metal/bin/python -m ruff check tests/test_sampling.py vllm_metal/v1/sampling_batch.py vllm_metal/envs.py
# All checks passed!

.venv-vllm-metal/bin/python -m ruff format --check tests/test_sampling.py vllm_metal/v1/sampling_batch.py vllm_metal/envs.py
# 3 files already formatted
```

## End-to-End Benchmark

This is not claimed as a performance win.

Benchmark setup:

- model: `Qwen/Qwen3-0.6B`
- revision: `c1899de289a04d12100db370d81485cdf75e47ca`
- path: paged attention
- env: `VLLM_METAL_USE_PAGED_ATTENTION=1`
- branch native env: `VLLM_METAL_NATIVE_RANDOM_SAMPLING=1`
- workload: random input 32 / output 128, 16 prompts, max concurrency 4
- sampling: `temperature=0.8`, `top_p=0.9`, `top_k=50`, `min_p=0.05`
- hardware: Apple M5, 16GB, macOS 26.4

3-run mean branch vs main:

| Metric | Result |
| --- | ---: |
| output throughput | -8.80% |
| total token throughput | -8.80% |
| mean TPOT | +16.87% |
| mean ITL | +34.95% |
| mean TTFT | +38.07% |

Given these results, the new path stays experimental and opt-in.

## Follow-up Work

Before this can be advertised as a performance improvement, the next steps are:

- profile the native path against the Torch bridge to identify whether the current cost comes from full-vocab sorting, extra MLX ops, or categorical sampling itself
- revisit the top-p pre-filter idea only if the candidate-set semantics can be kept exact
- evaluate larger-vocab and higher-concurrency decode-heavy workloads where sampler overhead is more visible
- add penalties and seeded RNG in separate PRs after their vLLM semantics are matched
- consider enabling the path by default only after a main-vs-branch end-to-end benchmark shows a stable win

### CODE REVIEW

**Review by WindChimeRan — COMMENTED** [review]
I run a sampling-only benchmark myself: 

A native sampler that's ~1.5× slower than the torch roundtrip it replaces (~6.6 vs ~4.4 ms/step) isn't acceptable even behind a flag

**Comment by LxYuan0420** [comment]
I am not convinced we should introduce a user-visible env var for a path that is currently worse than the default. I would prefer to keep this out of `main` until we have data showing where native MLX sampling can actually win. Otherwise, we add another sampler path to maintain without a clear user benefit today.

**Comment by CAICAIIs** [comment]
Thanks, that concern makes sense. Before pushing more changes, I did a local rework to make the path narrower and checked whether it has a real winning case.

In the local version, the native path stays default-off and is limited to no-logprobs paged decode for all-random batches. Logprobs, seeds/generators, penalties, allowed token filters, bad words, mixed greedy/random batches, custom processors, and prefill still fall back to the existing Torch sampler.

The local sampling-only repro is now faster than the Torch fallback:

- batch=4, vocab=151936, top_k=50, top_p=0.9, min_p=0.05: native 3.27 ms vs Torch fallback 5.08 ms
- batch=8, vocab=151936, top_p=0.9: native 4.25 ms vs Torch fallback 6.17 ms

I also ran a local main-vs-branch e2e A/B/A/B benchmark on a decode-heavy workload:

- model: Qwen/Qwen3-0.6B
- workload: random input 32, output 128, 16 prompts, concurrency 4
- sampling: temperature=0.8, top_p=0.9, top_k=50, min_p=0.05
- env: VLLM_METAL_USE_PAGED_ATTENTION=1, branch only VLLM_METAL_NATIVE_RANDOM_SAMPLING=1

| metric | main | local branch native | delta |
|---|---:|---:|---:|
| output tok/s | 163.249 ± 0.081 | 177.135 ± 0.237 | +8.51% |
| mean TPOT | 23.208 ± 0.069 ms | 21.441 ± 0.037 ms | -7.61% |
| p90 TPOT | 23.464 ± 0.061 ms | 21.579 ± 0.159 ms | -8.03% |

Would this narrower default-off direction, with the local e2e positive case above, be worth pushing for review? Or would you prefer to keep native random sampling out until there is broader workload coverage?

**Comment by LxYuan0420** [comment]
Thanks for exploring this. I think we should close this PR for now. Sampling is correctness-sensitive, and a second sampler path behind an env var is not worth carrying without a much stronger end-to-end performance case.

**Comment by CAICAIIs** [comment]
> Thanks for exploring this. I think we should close this PR for now. Sampling is correctness-sensitive, and a second sampler path behind an env var is not worth carrying without a much stronger end-to-end performance case.

got it, you are right
